dask_planner/src/sql/optimizer/filter_columns_post_join.rs

```diff
@@ -0,0 +1,678 @@
+                    // Recurse on inputs vector of Union
+                    let mut new_inputs = vec![];
+                    for input in &u.inputs {
+                        let new_input = optimize_top_down(
+                            input,
+                            &input.all_schemas(),
+                            post_join_columns.clone(),
+                        );
+                        match new_input {
+                            Ok(i) => new_inputs.push(Arc::new(i)),
+                            _ => {
+                                warn!("Skipping optimizer rule 'FilterColumnsPostJoin'");
+                                return Ok(plan.clone());
+                            }
+                        }
+                    }
+
+                    let union_plan = LogicalPlan::Union(Union {
+                        inputs: new_inputs,
+                        schema: u.schema.clone(),
+                        alias: u.alias.clone(),
+                    });
+
+                    // Add Union to HashMap
+                    new_plan.insert(counter, union_plan);
+                }
+                _ => {
+                    warn!("Skipping optimizer rule 'FilterColumnsPostJoin'");
+                    return Ok(plan.clone());
+                }
+            }
+            counter += 1;
+            break;
+        } else {
+            match current_plan {
+                LogicalPlan::Projection(ref p) => {
+                    new_plan.insert(counter, current_plan.clone());
+                    previous_projection = true;
+
+                    // Reset HashSet with projected columns
+                    post_join_columns = HashSet::new();
+                    insert_post_join_columns(&mut post_join_columns, &p.expr);
+                }
+                _ => {
+                    // Is not a Join, so just add LogicalPlan to HashMap
+                    new_plan.insert(counter, current_plan.clone());
+                    previous_projection = false;
+
+                    let current_columns = &current_plan.expressions();
+                    insert_post_join_columns(&mut post_join_columns, current_columns);
+                }
+            }
+            // Move on to next step
+            current_plan = current_plan.inputs()[0].clone();
+            counter += 1;
+        }
+    }
+
+    if join_flag {
+        counter -= 1;
+        // Organize HashMap into LogicalPlan to return
+        let mut previous_step = new_plan.get(&counter).unwrap();
+        let mut next_step;
+        let mut return_plan = plan.clone();
+
+        for key in (0..counter).rev() {
+            next_step = new_plan.get(&key).unwrap();
+            match next_step {
+                LogicalPlan::Projection(p) => {
+                    return_plan = LogicalPlan::Projection(Projection {
+                        expr: p.expr.clone(),
+                        input: Arc::new(previous_step.clone()),
```

Comment:
Not sure if there's a better way to rebuild the plan? Essentially we need to replace the `input` (an `Arc<LogicalPlan>`) of every `LogicalPlan` enum in the plan. So we're iterating through the plan and matching it with an enum, which requires checking if it is any of the possible enums (that have an `input` parameter) in https://github.com/apache/arrow-datafusion/blob/master/datafusion/expr/src/logical_plan/plan.rs#L48

Reply:
Still don't like this. It's essentially just 100 lines of code to check if there's an `input` that needs to be replaced.